### PR TITLE
EventSourcing stage not completed if no event has been emitted

### PR DESCRIPTION
**Pull request by odd**

### CODE REVIEW

**Comment by krasserm** [comment]
@odd thanks for your contribution! What's wrong with the existing code i.e. what does your change attempt to fix?

**Comment by odd** [comment]
Hi, this PR was a bit premature I'm afraid. I have since submitting it found that the change (while solving my initial problem with the initial command not succeeding as the event is never sent as a response after being emitted) instead causes all following commands to fail. 

Here is the code I'm testing:
```scala
  trait Aggregate[A] {
    def aggregateId(a: A): String
  }

  implicit val blogCommandAggregate: Aggregate[BlogCommand] = {
    case GetPost(id) => id.toString
    case AddPost(content) => PostId(content.title).toString
    case UpdatePost(id, _) => id.toString
  }

  def commandProcessor(aggregateId: String): Flow[BlogCommand, BlogResponse, NotUsed] =
    EventSourcing(aggregateId, BlogState(), requestHandler, eventHandler).join(akkaPersistenceEventLog.flow(aggregateId)).log("processor")

  def commandRouter(implicit agg: Aggregate[BlogCommand]): Flow[BlogCommand, BlogResponse, NotUsed] =
    Router(req => agg.aggregateId(req), (aggregateId: String) => commandProcessor(aggregateId))

  def handleCommand(cmd: BlogCommand) = Source.single(cmd).via(commandRouter).runWith(Sink.head)
```
The Akka Http route outside looks like this:
```scala
complete(handleCommand(cmd))
```

**Comment by odd** [comment]
The underlying problem seems to be that the `requestUpstreamFinished` flag becomes set to `false` too early for the initial command sent so that the `tryPullEi()` call in the `InHandler` does not pull any event. I have tested to make it a counter (`requestUpstreamFinishedState`) instead and used `requestUpstreamFinishedState <= 1` in `tryPullEI()` and instead used `requestUpstreamFinishedState > 0` in `tryPullCi()` and that seems to work for my use case (both for the initial command and any succeeding commands).

I'll try to create a self contained minimal use case demonstrating the problem (as soon as I can find the time)

**Comment by krasserm** [comment]
> I'll try to create a self contained minimal use case demonstrating the problem (as soon as I can find the time

That would be very helpful. I'm still on vacation this week. Will take a closer look next week. Cheers!

**Comment by odd** [comment]
All test (including the ones added in f61e9f1) now pass (since ae4f1fc).

File: src/main/scala/com/github/krasserm/ases/EventSourcing.scala [hunk 1]
```diff
@@ -227,9 +228,9 @@ private class EventSourcing[S, E, REQ, RES](
         tryPullEi()
 
       private def tryPullEi(): Unit =
-        if (!requestUpstreamFinished) pull(ei)
+        if (!requestUpstreamFinished && !hasBeenPulled(ei) && !isClosed(ei)) pull(ei)
```
Comment (krasserm) [inline src/main/scala/com/github/krasserm/ases/EventSourcing.scala]: I think this change is not needed. `tryPullEi` is called `onPush` within `ei` handler, hence `!isClosed` and `!hasBeenPulled` always evaluate to `true` and can therefore be removed.

File: src/main/scala/com/github/krasserm/ases/EventSourcing.scala [hunk 2]
```diff
@@ -227,9 +228,9 @@ private class EventSourcing[S, E, REQ, RES](
...
 
       private def tryPullCi(): Unit =
-        if (isAvailable(eo) && isAvailable(ro) && !hasBeenPulled(ci) && roundtrip.isEmpty && !requestUpstreamFinished) pull(ci)
+        if (!requestUpstreamFinished && recovered && roundtrip.isEmpty && isAvailable(eo) && isAvailable(ro) && !hasBeenPulled(ci) && !isClosed(ci)) pull(ci)
```
Comment (krasserm) [inline src/main/scala/com/github/krasserm/ases/EventSourcing.scala]: `!isClosed(ci)` is redundant with `!requestUpstreamFinished` and can be removed.

File: src/test/scala/com/github/krasserm/ases/EventSourcingSpec.scala [hunk 3]
```diff
@@ -32,17 +32,35 @@ object EventSourcingSpec {
...
+    override def delta = 0
+  }
   case class Response(state: Int)
 
-  val requestHandler: RequestHandler[Int, Incremented, Request, Response] = {
-    case (s, GetState)     => respond(Response(s))
-    case (_, Increment(d)) => emit(Seq(Incremented(d)), Response)
+  val requestHandler: RequestHandler[Int, Event, Request, Response] = {
+    case (s, req @ GetState)      =>
+      println(s"($s, $req)")
```
Comment (krasserm) [inline src/test/scala/com/github/krasserm/ases/EventSourcingSpec.scala]: Please, remove all `println`s.

File: src/test/scala/com/github/krasserm/ases/EventSourcingSpec.scala [hunk 4]
```diff
@@ -32,17 +32,35 @@ object EventSourcingSpec {
...
+      respond(Response(s))
+    case (s, req @ Increment(d))  =>
+      println(s"($s, $req)")
+      emit(Seq(Incremented(d)), Response)
+    case (s, req @ ClearIfEqualTo(v))    =>
+      println(s"($s, $req)")
+      if (s == v) emit(Seq(Cleared(v)), Response)
+      else emit(Seq(Kept(v)), Response)
```
Comment (krasserm) [inline src/test/scala/com/github/krasserm/ases/EventSourcingSpec.scala]: Replace with `respond(Response(s))` and remove definition of `Kept` event?

**Review by krasserm — APPROVED** [review]
Awesome, thanks! Please squash your commits

**Pull request merged**